windows/src/developer/TIKE/debug/Keyman.System.Debug.DebugEvent.pas

```diff
@@ -0,0 +1,393 @@
+    procedure SetEventType(const Value: TDebugEventType);
+  public
+    constructor Create;
+    destructor Destroy; override;
+    property Action: TDebugEventActionData read FAction;
+    property Rule: TDebugEventRuleData read FRule;
+    property Cursor: TDebugEventCursor read FCursor;
+    property EventType: TDebugEventType read FEventType write SetEventType;
+  end;
+
+  TDebugEventList = class(TObjectList<TDebugEvent>)
+  private
+    procedure Action_Char(const character: km_kbp_usv);
+    procedure Action_DeleteBack(expected_type: uint8_t;
+      expected_value: uintptr_t);
+    procedure Action_EmitKeystroke(const key: Word);
+    procedure Action_Marker(marker: uintptr_t);
+    function AddActionItem(key: Word; action: pkm_kbp_action_item): Boolean;
+    procedure AddDebugItem(
+      debug: pkm_kbp_state_debug_item;
+      debugkeyboard: TDebugKeyboard;
+      vk: uint16_t;
+      modifier_state: uint16_t
+    );
+  public
+    function AddStateItems(
+      state: pkm_kbp_state;
+      vk: uint16_t;
+      modifier_state: uint16_t;
+      debugkeyboard: TDebugKeyboard
+    ): Boolean; overload;
+
+    function AddStateItems(
+      state: pkm_kbp_state;
+      vk: uint16_t;
+      modifier_state: uint16_t
+    ): Boolean; overload;
+  end;
+
+implementation
+
+uses
+  kmxfile,
+  Unicode;
+
+{ TDebugEvent }
+
+constructor TDebugEvent.Create;
+begin
+  inherited Create;
+  FEventType := etAction;
+  FAction := TDebugEventActionData.Create;
+end;
+
+destructor TDebugEvent.Destroy;
+begin
+  FreeAndNil(FAction);
+  FreeAndNil(FRule);
+  inherited;
+end;
+
+procedure TDebugEvent.SetEventType(const Value: TDebugEventType);
+begin
+  if FEventType <> Value then
+  begin
+    case FEventType of
+      etAction:    FreeAndNil(FAction);
+      etRuleMatch: FreeAndNil(FRule);
+    end;
+    FEventType := Value;
+    case FEventType of
+      etAction:    FAction := TDebugEventActionData.Create;
+      etRuleMatch: FRule := TDebugEventRuleData.Create;
+    end;
+  end;
+end;
+
+function TDebugEventList.AddActionItem(key: Word; action: pkm_kbp_action_item): Boolean;
+begin
+  Result := True;
+  case action._type of
+    KM_KBP_IT_CHAR:           Action_Char(action.character);
+    KM_KBP_IT_MARKER:         Action_Marker(action.marker); // Correlates to kmn's "deadkey" markers.
+    KM_KBP_IT_ALERT:          ; // TODO: The keyboard has triggered a alert/beep/bell.
+    KM_KBP_IT_BACK:           Action_DeleteBack(action.backspace.expected_type, action.backspace.expected_value);
+    KM_KBP_IT_PERSIST_OPT:    ; // TODO: The indicated option needs to be stored.
+    KM_KBP_IT_EMIT_KEYSTROKE: Action_EmitKeystroke(key);
+    else Assert(False, 'Action type '+IntToStr(Ord(action._type))+' is unexpected.');
+  end;
+end;
+
+procedure TDebugEventList.Action_EmitKeystroke(const key: Word);
+var
+  event: TDebugEvent;
+begin
+  case key of
+    VK_TAB:    Action_Char(9);  // Emit a tab character
+    VK_RETURN: Action_Char(13); // New line character
```

Comment:
Magic numbers but at least the comment tells save you looking it up

Reply:
Inclined to let these ones slide given they are chars and not virtual keys (Delphi has no `\t`, instead it uses `#9`, so here I could do `Ord(#9)` for Tab but it makes no difference).